### DISCUSSION TITLE
Twice RTC drive

### DISCUSSION THREAD

**Discussion by mauroForlimpopoli** [Moddable-OpenSource/moddable#1074] [Ideas]
Good morning,
in my project I'm using RTC DS1307 and the driver works correctly. Sometimes I wish to replace the DS1307 with MCP79400I. It has the same PINOUT, and the location of the data memory is almost the same.
The only big difference is the address
const DS1307_ADDR = 0x68;
How can I change it dynamically on the DS1307.js file?
I thought something like 
```
export const newADD = 0x6F;
Object.freeze(newADD ,true);
```
in the main.js, but maybe is the wrong way and I can't how to import from DS1307.js
If possible I don't want to duplicate the drive...

How can I do ?
Regards

**Comment by phoddie** [comment]
The example at [drivers/peripherals/rtc/showTime](https://github.com/Moddable-OpenSource/moddable/blob/public/examples/drivers/peripherals/rtc/showTime/main.js) shows instantiation of the DS1307 driver as follows:

```js
const rtc = new RTC({
            clock: {
                ...device.I2C.default,
                io: device.io.SMBus
            }
});
```

The `clock` object is an ECMA-419 [I2C constructor options object](https://419.ecma-international.org/#-10-io-classes-ic). It configures the I2C communication, including the address. The DS1307 driver [defaults to address 0x68,](https://github.com/Moddable-OpenSource/moddable/blob/a06dd7fe30389887a77ce08e813d3ce39430f782/modules/drivers/peripherals/ds1307/ds1307.js#L43) as you note. The contents of the `clock` object can override that default. Therefore, you can override the default address by including the address property in the clock object:

```js
const rtc = new RTC({
            clock: {
                ...device.I2C.default,
                io: device.io.SMBus,
                address: 0x6F
            }
});
```

> Sometimes I wish to replace the DS1307 with MCP79400I. It has the same PINOUT, and the location of the data memory is almost the same

FWIW – I cannot say if this is safe. There are differences. But, perhaps if you are only using a subset of features (e.g. get and set time) it will work. Please test carefully to confirm that it is a safe assumption for production.

**Reply by mauroForlimpopoli** [reply]
Like you correctly said there are some difference that I considered and modified in the original driver.
Thank you for your assistance

**Reply by phoddie** [reply]
Great. If your question has been addressed, we should close out this discussion as resolved.

**Reply by mauroForlimpopoli** [reply]
Peter,
I was using the driver linked
"import RTC from "ds1307";"
and not 
"import RTC from "embedded:RTC/DS1307";"

In the old driver the change is not possible

So I tried to change path but I'm getting
```
# Break: (host): module "embedded:RTC/DS1307" not found!
```

These are the nodule about i2c loaded:
		"$(MODDABLE)/modules/drivers/peripherals/ds1307/manifest.json",
		"$(MODDABLE)/modules/drivers/rtc/manifest.json",
		"$(MODULES)/pins/i2c/manifest.json"

Do I need to change something in the manifest.json?
regards

**Reply by mauroForlimpopoli** [reply]
Peter, 
I found the i2c Scan function and I put it inside the constructor so I solved the problem.
the address are 0x68 and 0x6F
```
		function scan() {
			
			let address = 0x68
			while (address < 0x70) {
		
				const i2c = new I2C({address, timeout: 50});
				let found = undefined !== i2c.read(1);
				i2c.close();
		
				if (found) {
					myFound = address
					trace(`Found 0x${address.toString(16).padStart(2, "0")}\n`);
				}
				address = address + 7
			}
		}
		
		scan(); 
		super(Object.assign({address: myFound}, dictionary));
```
thank you Peter

**Reply by phoddie** [reply]
Glad you found a solution.
